lib/view_component/base.rb

```diff
@@ -293,6 +351,65 @@ def with_content_areas(*areas)
         self.content_areas = areas
       end
 
+      # support initalizing slots as:
+      #
+      # with_slot(
+      #   :header,
+      #   collection: true|false,
+      #   class_name: "Header" # class name string, used to instantiate Slot
+      # )
+      def with_slot(*slot_names, collection: false, class_name: nil)
+        slot_names.each do |slot_name|
+          # Ensure slot_name is not already declared
+          if self.slots.key?(slot_name)
+            raise ArgumentError.new("#{slot_name} slot declared multiple times")
+          end
+
+          # Ensure slot name is not :content
+          if slot_name == :content
+            raise ArgumentError.new ":content is a reserved slot name. Please use another name, such as ':body'"
+          end
+
+          # Set the name of the method used to access the Slot(s)
+          accessor_name =
+            if collection
+              # If Slot is a collection, set the accessor
+              # name to the pluralized form of the slot name
+              # For example: :tab => :tabs
+              ActiveSupport::Inflector.pluralize(slot_name)
+            else
+              slot_name
+            end
+
+          instance_variable_name = "@#{accessor_name}"
+
+          # If the slot is a collection, define an accesor that defaults to an empty array
+          if collection
+            class_eval <<-RUBY
+              def #{accessor_name}
+                #{instance_variable_name} ||= []
+              end
+            RUBY
+          else
+            attr_reader accessor_name
+          end
+
+          # Generate a Slot class unless one is provided.
+          # with_slot(:header) generates MyComponent::Header < ViewComponent::Slot
+          unless class_name.present?
+            self.const_set(slot_name.to_s.capitalize, Class.new(ViewComponent::Slot))
+            class_name = "#{self}::#{slot_name.to_s.capitalize}"
+          end
```

Comment:
Should we raise if slot doesn't inherit from `ViewComponent::Slot`?

Reply:
@BlakeWilliams this is constructing a dynamic class, and only gets called if no class name is specified!

Reply:
@bbugh Right, if it's not present. I tried putting the comment below to not associate it with the auto-generation but I don't think I did a good job of getting that across. :)

If we pass in a `class_name` argument that doesn't inherit from `ActionView::Slot`, maybe it's worth raising instead of relying on behavior failing? It would definitely be a development/help oriented raise + error message imo.

Reply:
Thanks for explaining! I agree that sounds useful, great catch.

Reply:
Resolved.